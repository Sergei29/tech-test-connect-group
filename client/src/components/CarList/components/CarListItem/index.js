export { default } from "./CarListItem.component";
