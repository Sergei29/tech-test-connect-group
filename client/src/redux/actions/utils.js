/**
 * @description generates the error message out of received error from api
 * @param {Object} error error received from api
 * @param {String} entity name of the entity: 'carList' or 'carCurrent'
 * @returns {String} aggregated error message
 */
export const getErrorMessage = (error, entity) => {
  let errorMessage = "API call failed.";

  if (error.response && error.response.data) {
    errorMessage = error.response.data;
  } else if (error.message) {
    errorMessage = error.message;
  }

  return `${errorMessage}. ${entity}`;
};
