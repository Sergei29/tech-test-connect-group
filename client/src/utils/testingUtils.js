import React from "react";
import { shallow, mount } from "enzyme";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);

const store = mockStore({
  carList: [
    {
      id: "xe",
      modelYear: "k17",
      url: "/vehicle/xe",
      description:
        "The most advanced, efficient and refined sports saloon that Jaguar has ever produced",
      price: "£30,000",
      media: [{ media: "vehicle", url: "assets/images/xe_k17.jpg" }],
    },
    {
      id: "xf",
      modelYear: "k17",
      url: "/vehicle/xf",
      description:
        "Luxury business saloon with distinctive design, dynamic drive and state-of-the-art technologies.",
      price: "£36,000",
      media: [{ media: "vehicle", url: "assets/images/xf_k17.jpg" }],
    },
  ],
  carCurrent: {
    id: "xe",
    meta: {
      passengers: 5,
      drivetrain: ["AWD", "RWD"],
      bodystyles: ["saloon"],
      emissions: { template: "CO2 Emissions $value g/km", value: 99 },
    },
  },
  error: null,
  loading: false,
});

export const shallowRendererReduxConnected = (ComponentName, props = {}) =>
  shallow(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const mountRendererReduxConnected = (ComponentName, props = {}) =>
  mount(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const shallowRenderer = (ComponentName, props = {}) =>
  shallow(<ComponentName {...props} />);

export const mountRenderer = (ComponentName, props = {}) =>
  mount(
    <BrowserRouter>
      <ComponentName {...props} />
    </BrowserRouter>
  );
