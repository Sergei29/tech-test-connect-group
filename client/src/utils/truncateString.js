export const truncateString = (stringValue, maxLength) => {
  let truncateLocation = stringValue.indexOf(" ", maxLength);
  if (truncateLocation === -1) return stringValue;
  return stringValue.substring(0, truncateLocation) + "...";
};
