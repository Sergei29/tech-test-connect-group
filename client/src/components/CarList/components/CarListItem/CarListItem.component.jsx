import React from "react";
import PropTypes from "prop-types";
import { truncateString } from "../../../../utils/truncateString";
//styles:
import { CarListItemContainer } from "./CarListItem.style";

/**
 * @description functional component for car list item
 * @param {String} {carName car mark model
 * @param {String} imgUrl path to image
 * @param {String} carDetailsUrl path to car detals page
 * @param {String} carSummary brief description
 * @param {String} carPrice} price
 * @returns {JSX} component markup
 */
const CarListItem = ({
  carName,
  imgUrl,
  carDetailsUrl,
  carSummary,
  carPrice,
}) => {
  return (
    <CarListItemContainer to={carDetailsUrl}>
      <div className="listItem__image_box">
        <img src={imgUrl} alt={`vehicle ${carName}`} />
      </div>
      <h3 className="listItem__car_name">
        <span>{carName}</span>
      </h3>
      <p className="listItem__car_price">{carPrice}</p>
      <p className="listItem__car_summary">{truncateString(carSummary, 50)}</p>
    </CarListItemContainer>
  );
};

CarListItem.propTypes = {
  carName: PropTypes.string.isRequired,
  imgUrl: PropTypes.string.isRequired,
  carDetailsUrl: PropTypes.string.isRequired,
  carSummary: PropTypes.string.isRequired,
  carPrice: PropTypes.string.isRequired,
};

export default CarListItem;
