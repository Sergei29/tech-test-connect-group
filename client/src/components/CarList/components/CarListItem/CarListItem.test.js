import { shallowRenderer } from "../../../../utils/testingUtils";
import CarListItem from "./CarListItem.component";

describe("CarListItem", () => {
  let wrapper, props;

  beforeEach(() => {
    props = {
      carName: "Jaguar k17 xe",
      imgUrl: "assets/images/xe_k17.jpg",
      carDetailsUrl: "/vehicle/xe",
      carSummary:
        "The most advanced, efficient and refined sports saloon that Jaguar has ever produced",
      carPrice: "Â£30,000",
    };
    wrapper = shallowRenderer(CarListItem, props);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
