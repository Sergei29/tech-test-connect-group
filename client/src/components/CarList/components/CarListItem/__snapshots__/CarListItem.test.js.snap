// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarListItem should match the snapshot 1`] = `
<Styled(Link)
  to="/vehicle/xe"
>
  <div
    className="listItem__image_box"
  >
    <img
      alt="vehicle Jaguar k17 xe"
      src="assets/images/xe_k17.jpg"
    />
  </div>
  <h3
    className="listItem__car_name"
  >
    <span>
      Jaguar k17 xe
    </span>
  </h3>
  <p
    className="listItem__car_price"
  >
    Â£30,000
  </p>
  <p
    className="listItem__car_summary"
  >
    The most advanced, efficient and refined sports saloon...
  </p>
</Styled(Link)>
`;
